### PR TITLE
Improve dotCover get\set coverage report

### LINKED ISSUE #3254: dotCover - improve getter/setter partial coverage support
See #3252 and #2622

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**
Fixes #3254

### CODE REVIEW

File: sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/DotCoverReportParser.java [hunk 1]
```diff
@@ -36,7 +41,7 @@
   private static final Pattern COVERED_LINES_PATTERN_1 = Pattern.compile(
     ".*<script type=\"text/javascript\">\\s*+highlightRanges\\(\\[(.*?)\\]\\);\\s*+</script>.*",
     Pattern.DOTALL);
-  private static final Pattern COVERED_LINES_PATTERN_2 = Pattern.compile("\\[(\\d++),\\d++,\\d++,\\d++,(\\d++)\\]");
+  private static final Pattern COVERED_LINES_PATTERN_2 = Pattern.compile("\\[(\\d++),\\d++,(\\d++),\\d++,(\\d++)\\]");
```
Comment (andrei-epure-sonarsource) [inline sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/DotCoverReportParser.java]: why was this change needed?

File: sonar-dotnet-shared-library/src/test/resources/dotcover/valid_multiple_sequence_points_per_line.html [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+<html>
+  <head>
+    <meta http-equiv="content-type" content="text/html; charset=utf-8" />
+    <title>GetSet\Bar.cs</title>
+    <script type="text/javascript" src="../js/dotcover.sourceview.js"></script>
+    <link rel="stylesheet" type="text/css" href="../css/dotcover.report.css" />
+  </head>
+  <body>
+    <pre id="content" class="source-code">
```
Comment (andrei-epure-sonarsource) [inline sonar-dotnet-shared-library/src/test/resources/dotcover/valid_multiple_sequence_points_per_line.html]: nice!

File: sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/DotCoverReportParser.java [hunk 3]
```diff
@@ -100,13 +105,34 @@ private void collectCoverage(String fileCanonicalPath, String contents) {
...
-        int line = Integer.parseInt(matcher.group(1));
-        int hits = Integer.parseInt(matcher.group(2));
-        coverage.addHits(fileCanonicalPath, line, hits);
+        int lineStart = Integer.parseInt(matcher.group(1));
+        int lineEnd = Integer.parseInt(matcher.group(2));
+        int hits = Integer.parseInt(matcher.group(3));
+
+        coverage.addHits(fileCanonicalPath, lineStart, hits);
+
+        if (lineStart == lineEnd) {
```
Comment (andrei-epure-sonarsource) [inline sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/DotCoverReportParser.java]: should you add a comment to explain why only if `lineStart==lineEnd`?

File: sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/DotCoverReportParser.java [hunk 4]
```diff
@@ -100,13 +105,34 @@ private void collectCoverage(String fileCanonicalPath, String contents) {
...
+          sequencePoints.add(new SequencePoint(lineStart, hits));
+        }
 
         LOG.trace("dotCover parser: found coverage for line '{}', hits '{}' when analyzing the path '{}'.",
-            line, hits, fileCanonicalPath);
+            lineStart, hits, fileCanonicalPath);
+      }
+
+      Map<Integer, List<SequencePoint>> collect = sequencePoints.stream().collect(Collectors.groupingBy(SequencePoint::getStartLine));
```
Comment (andrei-epure-sonarsource) [inline sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/DotCoverReportParser.java]: s/collect/sequencePointsPerLine ?

File: sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/DotCoverReportParser.java (hunk 1 above)
Comment (costin-zaharia-sonarsource) [reply sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/DotCoverReportParser.java]: In order to be able to extract the endLine with the matcher

File: sonar-dotnet-shared-library/src/test/resources/dotcover/valid_multiple_sequence_points_per_line.html (hunk 2 above)
Comment (costin-zaharia-sonarsource) [reply sonar-dotnet-shared-library/src/test/resources/dotcover/valid_multiple_sequence_points_per_line.html]: it's the dotCover format 😄

File: sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/DotCoverReportParser.java (hunk 4 above)
Comment (costin-zaharia-sonarsource) [reply sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/DotCoverReportParser.java]: Yes. I thought to change it and then forgot 😄
